### PR TITLE
Speeding up optspec and meta.record_ypos

### LINKED ISSUE #408: Recent drastic slowdown in optimal spectral extraction
### Instrument

MIRI (Stages 1-3)

### What happened?

In the past week, my optimal extraction times for the MIRI commissioning data have inexplicably gone from ~1 minute (2.8 minutes for all of S3) to ~1h20m when the exact same ECF (with ncpu=14, isplots_S3=1)... Not sure what could have gone wrong here, but one of the recent PRs must have drastically slowed things unintentionally

### Error traceback output

_No response_

### What operating system are you using?

_No response_

### What version of Python are you running?

_No response_

### What Python packages do you have installed?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by taylorbell57**
Using parallelized code to speed up meta.record_ypos which had been
slow. Also moved code from s3_reduce to source_pos and optspex to
better modularize the code.

Also added the ability to set a limit on how many figures are made of a
certain type (currently only applies to Figs 3301, 3302, 3304, and 3501
which are the most annoying figures to get tonnes of.

Resolves #408

### CODE REVIEW

File: src/eureka/S3_data_reduction/nirspec.py [hunk 1]
```diff
@@ -280,15 +280,16 @@ def straighten_trace(data, meta, log):
     meta : eureka.lib.readECF.MetaClass
         The updated metadata object.
     '''
-    log.writelog('  Correcting curvature and bringing trace in the center '
-                 'of the detector...', mute=(not meta.verbose))
+    log.writelog('  Correcting G395H curvature and bringing the trace to the '
```
Comment (kevin218) [inline src/eureka/S3_data_reduction/nirspec.py:283]: NIRCam also has some curvature, so let's now specifically call out G395. Also, G235 may have curvature.

File: src/eureka/S3_data_reduction/nirspec.py [hunk 2]
```diff
@@ -320,11 +321,6 @@ def straighten_trace(data, meta, log):
                  mute=(not meta.verbose))
     meta.src_ypos = new_center
 
-    # update the median frame
```
Comment (kevin218) [inline src/eureka/S3_data_reduction/nirspec.py:323]: I don't think you want to delete these lines. We need to recalculate the median frame.

File: src/eureka/S3_data_reduction/optspex.py [hunk 3]
```diff
@@ -477,9 +478,99 @@ def profile_gauss(subdata, mask, threshold=10, guess=None, isplots=0):
     return profile
 
 
-def optimize(meta, subdata, mask, bg, spectrum, Q, v0, p5thresh=10,
-             p7thresh=10, fittype='smooth', window_len=21, deg=3,
-             windowtype='hanning', n=0, m=0, meddata=None):
+def optimize(data, meta, log, apdata, apmask, apbg, apv0, gain=1,
```
Comment (kevin218) [inline src/eureka/S3_data_reduction/optspex.py:481]: I would call this routine 'optimize_wrapper' and leave optimize as previously defined.

**Comment by kevin218** [comment]
With this PR, can you also make 'driftypos' and 'driftywidth' level 1 plots, instead of level 3?

File: src/eureka/S3_data_reduction/source_pos.py [hunk 4]
```diff
@@ -2,10 +2,12 @@
 
 import numpy as np
 from scipy.optimize import curve_fit
+from tqdm import tqdm
+import multiprocessing as mp
 from . import plots_s3
 
 
-def source_pos(data, meta, m, integ=0):
+def source_pos(data, meta, log, m, integ=0):
```
Comment (kevin218) [inline src/eureka/S3_data_reduction/source_pos.py:10]: Rename to source_pos_wrapper

File: src/eureka/S3_data_reduction/source_pos.py [hunk 5]
```diff
@@ -101,36 +203,37 @@ def source_pos_max(flux, meta, m, integ=0, plot=True):
...
+    x_dim = flux.shape[0]
 
-    x_dim = flux.shape[1]
-
-    sum_row = np.ma.sum(flux[integ], axis=1)
+    sum_row = np.ma.sum(flux, axis=1)
     pos_max = np.ma.argmax(sum_row)
 
     # Diagnostic plot
     if meta.isplots_S3 >= 3 and plot:
```
Comment (kevin218) [inline src/eureka/S3_data_reduction/source_pos.py:212]: Change to >= 1 and rename figures to 31..

File: src/eureka/S3_data_reduction/source_pos.py [hunk 6]
```diff
@@ -148,22 +251,21 @@ def source_pos_FWM(flux, meta, m, integ=0):
...
+                                      pos_max+meta.spec_hw]
     sum_row -= (sum_row[0]+sum_row[-1])/2
 
-    y_pos = np.ma.sum(sum_row * y_pixels)/np.ma.sum(sum_row)
+    y_pos = np.ma.sum(sum_row*y_pixels)/np.ma.sum(sum_row)
 
     # Diagnostic plot
-    if meta.isplots_S3 >= 3:
-        plots_s3.source_position(meta, x_dim, pos_max, m, isFWM=True,
+    if meta.isplots_S3 >= 3 and plot:
```
Comment (kevin218) [inline src/eureka/S3_data_reduction/source_pos.py:267]: >= 1

File: src/eureka/S3_data_reduction/s3_reduce.py [hunk 7]
```diff
@@ -291,18 +286,10 @@ def reduce(eventlabel, ecf_path=None, s2_meta=None):
...
-                    data.flux.attrs['flux_units']
-
                 # correct G395H curvature
-                if meta.inst == 'nirspec' and data.mhdr['GRATING'] == 'G395H':
-                    if meta.curvature == 'correct':
-                        log.writelog('  Correcting for G395H curvature...',
-                                     mute=(not meta.verbose))
-                        data, meta = inst.straighten_trace(data, meta, log)
+                if (meta.inst == 'nirspec' and data.mhdr['GRATING'] == 'G395H'
+                        and meta.curvature == 'correct'):
```
Comment (kevin218) [inline src/eureka/S3_data_reduction/s3_reduce.py:291]: Generalize to allow other instruments or modes to correct curvature

**Review by kevin218 — CHANGES_REQUESTED** [review]

File: src/eureka/S3_data_reduction/nirspec.py (hunk 1 above)
Comment (taylorbell57) [reply src/eureka/S3_data_reduction/nirspec.py:283]: Do you mean *not* instead of now?

File: src/eureka/S3_data_reduction/nirspec.py (hunk 2 above)
Comment (taylorbell57) [reply src/eureka/S3_data_reduction/nirspec.py:323]: The median frame gets calculated in optspex later, so no need to do this twice

**Review by kevin218 — APPROVED** [review]

**Pull request merged**